### PR TITLE
chore: Add docs to monomorphization pass

### PR DESCRIPTION
**Pull request by jfecher**
# Description

Adds doc comments to various functions, structs, and modules in the monomorphization pass

### CODE REVIEW

File: crates/noirc_frontend/src/monomorphization/ast.rs [hunk 1]
```diff
@@ -120,6 +133,18 @@ pub struct Index {
     pub index: Box<Expression>,
 }
 
+/// Rather than a Pattern containing possibly several variables, Let now
+/// defines a single variable with the given LocalId. By the time this
+/// is produced in monomorphization, let-statements with tuple and struct patterns:
+/// ```nr
+/// let MyStruct { field1, field2 } = get_struct();
+/// ```
+/// Have been desugared into multiple let statements for simplicity:
```
Comment (TomAFrench) [inline crates/noirc_frontend/src/monomorphization/ast.rs:142]:
```suggestion
/// have been desugared into multiple let statements for simplicity:
```

File: crates/noirc_frontend/src/monomorphization/mod.rs [hunk 2]
```diff
@@ -1,3 +1,13 @@
+//! Coming after type checking, monomorphization is the last pass in Noir's frontend.
+//! It accepts the type checked HIR as input and produces a monomorphized AST as output.
+//! This file implements the pass itself, while the AST is defined in the ast module.
+//!
+//! Unlike the HIR which is stored within the NodeInterner, the monomorphized AST is
```
Comment (TomAFrench) [inline crates/noirc_frontend/src/monomorphization/mod.rs:5]:
```suggestion
//! Unlike the HIR, which is stored within the NodeInterner, the monomorphized AST is
```

File: crates/noirc_frontend/src/monomorphization/mod.rs [hunk 3]
```diff
@@ -18,16 +28,33 @@ use self::ast::{Definition, FuncId, Function, LocalId, Program};
 pub mod ast;
 pub mod printer;
 
+/// The context struct for the monomorphization pass.
+///
+/// This uses the NodeInterner internally to reference existing definitions in the HIR,
+/// but since the produced monomorphized AST uses fresh IDs and is self-contained, there
+/// is no more need for the NodeInterner after this pass.
```
Comment (TomAFrench) [inline crates/noirc_frontend/src/monomorphization/mod.rs:35]: I think we can move the first half of this to the `nodeInterner` field and remove the rest. We document that we don't use the `NodeInterner` after monomorphisation in a few places.

**Review by TomAFrench — COMMENTED** [review]
mostly nits, can check again tomorrow.

**Review by kevaundray — APPROVED** [review]

**Pull request merged**